### PR TITLE
Simplify *chown* syscalls

### PR DESCRIPTION
**Pull request by kripken**
This avoids musl calling the chown/lchown syscalls, and using fchown/fchownat
only. This leaves less work on the syscalls layer, so it will help WasmFS.

Also simplify the logic in fchown. For some reason musl tries chown on a path
of `/proc/fd/NNN` if fchown `NNN` fails. It seems like both should fail at equal times?
At least in our implementation of the filesystem they do, unless I'm missing
something.

Also add support for the nofollow flag to fchownat, which is now necessary.

### CODE REVIEW

File: system/lib/libc/musl/src/unistd/fchown.c [hunk 1]
```diff
@@ -10,8 +10,9 @@ int fchown(int fd, uid_t uid, gid_t gid)
 {
 	int ret = __syscall(SYS_fchown, fd, uid, gid);
 #if __EMSCRIPTEN__
-	if (ret != -EBADF || !__wasi_fd_is_valid(fd))
-		return __syscall_ret(ret);
+	// We can't continue onwards to try the /proc/fd/NNN approach that musl does,
+	// as we don't support that much of POSIX.
+	return __syscall_ret(ret);
 #else
```
Comment (sbc100) [inline system/lib/libc/musl/src/unistd/fchown.c:16]: I think the #endif here can now move all the way to the end of this function.

**Review by sbc100 — APPROVED** [review]
Great!

File: system/lib/libc/musl/arch/emscripten/bits/syscall.h [hunk 2]
```diff
@@ -53,7 +53,6 @@
 #define SYS_stat64		__syscall_stat64
 #define SYS_lstat64		__syscall_lstat64
 #define SYS_fstat64		__syscall_fstat64
-#define SYS_lchown32		__syscall_lchown32
```
Comment (sbc100) [inline system/lib/libc/musl/arch/emscripten/bits/syscall.h:56]: Can you also remove these from `system/lib/libc/musl/arch/emscripten/syscall_arch.h`?

**Review by tlively — APPROVED** [review]

**Review by sbc100 — APPROVED** [review]

File: src/library_syscall.js [hunk 3]
```diff
@@ -923,11 +913,13 @@ var SyscallsLibrary = {
...
 #endif
     path = SYSCALLS.getStr(path);
+    var nofollow = flags & {{{ cDefine('AT_SYMLINK_NOFOLLOW') }}};
+    flags = flags & (~{{{ cDefine('AT_SYMLINK_NOFOLLOW') }}});
 #if ASSERTIONS
     assert(flags === 0);
 #endif
     path = SYSCALLS.calculateAt(dirfd, path);
-    FS.chown(path, owner, group);
+    (nofollow ? FS.lchown : FS.chown)(path, owner, group);
```
Comment (sbc100) [inline src/library_syscall.js:922]: I wonder if its worth refactoring the FS api such that nofollow is a param there too?

**Review by sbc100 — APPROVED** [review]

File: src/library_syscall.js (hunk 3 above)
Comment (kripken) [reply src/library_syscall.js:922]: I think that is done internally:

https://github.com/emscripten-core/emscripten/blob/b2b97c22217e8b2e795f0aa5f82f3ca2ec529f74/src/library_fs.js#L927-L929

Perhaps the idea was to have the JS API mirror the POSIX API as much as possible, then internally it becomes a boolean. I am not sure though - that was many years ago and I didn't write the code. But I can see a good case both ways.

In any case as this is for the old FS, maybe best to leave it?

Comment (sbc100) [reply src/library_syscall.js:922]: Sure, sounds good.

**Pull request merged**